src/dune_engine/build_system.ml

```diff
@@ -2185,14 +2220,28 @@ let process_exn_and_reraise exn =
   in
   let build = get_build_system () in
   build.errors <- exn :: build.errors;
+  let+ () =
+    Memo.Build.of_reproducible_fiber (build.handler.error [ Handler.Add exn ])
+  in
   Exn_with_backtrace.reraise exn
 
 let run f =
   Hooks.End_of_build.once Promotion.finalize;
   let t = get_build_system () in
+  let open Fiber.O in
+  let* () =
+    t.handler.error (List.map t.errors ~f:(fun x -> Handler.Remove x))
+  in
+  t.errors <- [];
   let f () =
-    Memo.Build.run
-      (Memo.Build.with_error_handler f ~on_error:process_exn_and_reraise)
+    let* () = t.handler.build_event Start in
+    let open Fiber.O in
+    let* res =
+      Memo.Build.run
+        (Memo.Build.with_error_handler f ~on_error:process_exn_and_reraise)
+    in
+    let+ () = t.handler.build_event Finish in
```

Comment:
We never send this event if the build fails. Is that expected?

Reply:
Yes, that is expected. There supposed to be a `Fail` event in such a case. I will add it